File: CordovaApp/platforms/windows/www/plugins/com.msopentech.allJoyn/www/InterfaceDescription.js
cordova.define("com.msopentech.allJoyn.InterfaceDescription", function(require, exports, module) { 
// Copyright (c) Microsoft Open Technologies, Inc.  All rights reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

/*global module, require, msopentech*/

var ALLJOYN_PROXY = 'AllJoyn';

var exec = require('cordova/exec');
var checkArgs = require('cordova/argscheck').checkArgs;

/**
 * Class for describing message bus interfaces.
 * InterfaceDescription objects describe the methods, signals and properties of a BusObject or ProxyBusObject.
 * @param {[type]} name [description]
 */
var InterfaceDescription = function(name) {
    this.name = name;
    this._methods = [];
    this._properties = [];
};

/**
 * Add a method call member to the interface.
 * @param {String} name         Name of method call member.
 * @param {String} inSignature  Signature of input parameters or NULL for none.
 * @param {String} outSignature Signature of output parameters or NULL for none.
 * @param {String} argNames     Comma separated list of input and then output arg names used in annotation XML.
 * @param {Number} annotation   Annotation flags.
 */
InterfaceDescription.prototype.addMethod = function (name, inSignature, outSignature, argNames, annotation) {
    checkArgs('sSSSN', 'InterfaceDescription.addMethod', arguments);
    this._methods.push({
        name: name,
        inSig: inSignature || "",
        outSig: outSignature || "",
        argNames: argNames || "",
        annotation: annotation || 0
    });
};

/**
 * Add a property to the interface.
 * @param {String} name         Name of property.
 * @param {String} signature    Property type.
 * @param {Number} accessFlag   Access permission requirements on this call. One of
 *                                  msopentech.alljoyn.PROP_ACCESS_READ, msopentech.alljoyn.PROP_ACCESS_WRITE
 *                                  or msopentech.alljoyn.PROP_ACCESS_RW
 */
InterfaceDescription.prototype.addProperty = function (name, signature, accessFlag) {
    checkArgs('sSN', 'InterfaceDescription.addProperty', arguments);
    this._properties.push({
        name: name,
        sig: signature || "",
        access: accessFlag || msopentech.alljoyn.PROP_ACCESS_READ
    });
};

/**
 * Activate this interface. An interface must be activated before it can be used.
 * Activating an interface locks the interface so that is can no longer be modified.
 * @param  {Function} successCallback Success callback
 * @param  {Function} errorCallback   Error callback that accepts error code for this operation
 *                                        * ER_BUS_PROPERTY_ALREADY_EXISTS if the property can not be added because it already exists.
 *                                        * ER_BUS_MEMBER_ALREADY_EXISTS if member already exists
 */
InterfaceDescription.prototype.activate = function (successCallback, errorCallback) {
    checkArgs('fF', 'InterfaceDescription.activate', arguments);

    var that = this;

    var activate = function() {
        exec(successCallback, errorCallback, ALLJOYN_PROXY, 'interfaceDescriptionActivate', [that._id]);
    };

    if (that._methods.length > 0) {
        exec(function() {
            if (that._properties.length > 0) {
                exec(activate, errorCallback, ALLJOYN_PROXY, 'interfaceDescriptionAddProperties', [that._id, JSON.stringify(that._properties)]);
                return;
            }
            activate();
        }, errorCallback, ALLJOYN_PROXY, 'interfaceDescriptionAddMethods', [that._id, JSON.stringify(that._methods)]);
        return;
    }
    activate();
};

module.exports = InterfaceDescription;

});

File: CordovaApp/plugins/com.msopentech.allJoyn/src/windows/MSOpenTech.AllJoyn/InterfaceDescription.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All rights reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

namespace MSOpenTech.AllJoyn
{
    using System;
    using System.Runtime.InteropServices.WindowsRuntime;
    using System.Runtime.Serialization;

    using InterfaceDescription = AllJoynUnity.AllJoyn.InterfaceDescription;
    using QStatus = AllJoynUnity.AllJoyn.QStatus;

    // ReSharper disable once UnusedMember.Global
    public static partial class CordovaProxy
    {
        [DataContract]
        private struct MethodDescriptionContract
        {
            // ReSharper disable InconsistentNaming
            #pragma warning disable 649

            [DataMember]
            public string name;

            [DataMember]
            public string inSig;

            [DataMember]
            public string outSig;

            [DataMember]
            public string argNames;

            [DataMember]
            public InterfaceDescription.AnnotationFlags annotation;

            #pragma warning restore 649
        }

        [DataContract]
        private struct PropertyDescriptionContract
        {
            #pragma warning disable 649

            [DataMember]
            public string name;

            [DataMember]
            public string sig;

            [DataMember]
            public InterfaceDescription.AccessFlags access;

            #pragma warning restore 649
            // ReSharper restore InconsistentNaming
        }

        #region Proxy public methods
        // ReSharper disable UnusedMember.Global, InconsistentNaming
        
        public static void interfaceDescriptionAddMethods(CordovaSuccessCallback successCB, CordovaErrorCallback errorCB, [ReadOnlyArray] String[] args)
        {
            var interfaceDesc = args.GetInterfaceDescription();
            if (interfaceDesc == null)
            {
                errorCB(QStatus.BUS_OBJECT_NO_SUCH_INTERFACE);
                return;
            }
            
            var summaryResult = QStatus.OK;
            var methodDescriptions = Utils.Deserialize<MethodDescriptionContract[]>(args.GetArg(1));

            foreach (var m in methodDescriptions)
            {
                summaryResult = interfaceDesc.AddMethod(m.name, m.inSig, m.outSig, m.argNames, m.annotation);
            }

            if (summaryResult != QStatus.OK)
            {
                errorCB(summaryResult);
                return;
            }

            successCB(String.Empty);
        }

        public static void interfaceDescriptionAddProperties(CordovaSuccessCallback successCB, CordovaErrorCallback errorCB, [ReadOnlyArray] String[] args)
        {
            var interfaceDesc = args.GetInterfaceDescription();
            if (interfaceDesc == null)
            {
                errorCB(QStatus.BUS_OBJECT_NO_SUCH_INTERFACE);
                return;
            }

            var summaryResult = QStatus.OK;
            var propertyDescriptions = Utils.Deserialize<PropertyDescriptionContract[]>(args.GetArg(1));

            foreach (var p in propertyDescriptions)
            {
                summaryResult = interfaceDesc.AddProperty(p.name, p.sig, p.access);
            }

            if (summaryResult != QStatus.OK)
            {
                errorCB(summaryResult);
                return;
            }

            successCB(String.Empty);
        }

        public static void interfaceDescriptionActivate(CordovaSuccessCallback successCB, CordovaErrorCallback errorCB, [ReadOnlyArray] String[] args)
        {
            var interfaceDesc = args.GetInterfaceDescription();
            if (interfaceDesc == null)
            {
                errorCB(QStatus.BUS_OBJECT_NO_SUCH_INTERFACE);
                return;
            }

            interfaceDesc.Activate();
            successCB(String.Empty);
        }

        // ReSharper restore UnusedMember.Global, InconsistentNaming
        #endregion
    }
}
